063-topic-replication-factor-change.md

```diff
@@ -0,0 +1,271 @@
+  ```
+
+### Reconciliation logic
+
+At runtime, the Topic Operator watches and periodically reconciles the configuration of all managed and unpaused KafkaTopic resources.
+Among the other configurations, the reconciliation logic detects `KafkaTopic.spec.replicas` changes by comparing with the topic replication factor value in Kafka.
+
+A ReplicasChangeClient is used to isolate the logic required to call the Cruise Control REST API from the rest of the reconciliation logic.
+The ReplicasChangeClient supports sending multiple replicas changes at the same time.
+
+On every reconciliation loop, when Cruise Control integration is enabled and replicas changes are detected, the BatchingTopicController uses the ReplicasChangeClient for the following operations:
+
+1. Request pending replicas changes (uses `topic_configuration` endpoint, and returns topics with pending and ongoing changes).
+2. Check ongoing replicas changes (uses `user_tasks` endpoint, and returns topics with completed changes).
+
+The result of these operations is used to update the KafkaTopic status.
+
+### Error handling
+
+When there is an error, the replicas change remains pending or ongoing (no status update), and it is indefinitely retried by the periodic reconciliation.
+All errors are written in the Topic Operator logs, so that the user can verify the reason of a never ending replicas change.
+
+In Cruise Control, there is no way to differentiate between temporary errors (resolve by themself) from permanent errors (require some action).
+For example, the `NotEnoughValidWindowsException` can be raised when Cruise Control has just started and the cluster model is still building, but also when there is a configuration error which prevents broker metrics collection.
+Another example is the `OptimizationFailureException` raised when some hard goal is violated, which can be temporary if the configured network capacity is less than the real network capacity and there is a traffic peak.
+Instead, the `OngoingExecutionException` is clearly temporary and it means that there is already a task that is being executed by Cruise Control.
+
+Cruise Control allows to scale down the replication factor under the `min.insync.replicas` value, and this can cause disruption to producers with `acks=all`.
+When this happens, the Topic Operator logs a warning but continues with the replicas change, because the KafkaRoller ignores topics with RF < minISR, and they don't event show up as under replicated.
+The target replication factor should also be less than or equal to the number of available brokers, but this is enforced directly by Cruise Control.
+
+When a manged KafkaTopic with replicas change is deleted in Kubernetes, the Topic Operator also deletes the topic in Kafka, but the task execution continues in Cruise Control.
+There is little benefit in using the `stop_proposal_execution` endpoint, because most replicas change tasks will have only one batch that can't be stopped.
```

Comment:
So, what happens with such task? Will it just error in CC? I guess it is important to make sure it does not keep hanging there because of some race condition. Should we use the finalizers to prevent this from happening and delete the topic only once the task completes?

(to be clear, I'm fine with what you propose and not adding the complexity if we know for sure it will not have any negative side-effects)

Reply:
CC simply continues to execute the batch to the end. It's unnecessary work, because the topic is already deleted, but there is no harm (no task hang or error). Clarified in the proposal.